source/opt/inline_pass.cpp

```diff
@@ -0,0 +1,430 @@
+            SpvStorageClassFunction)
+      break;
+  }
+  return (type_itr != module_->types_values_end()) ? type_itr->result_id() : 0;
+}
+
+uint32_t InlinePass::AddPointerToType(uint32_t typeId) {
+  uint32_t resultId = TakeNextId();
+  std::vector<ir::Operand> in_operands;
+  in_operands.emplace_back(
+      spv_operand_type_t::SPV_OPERAND_TYPE_STORAGE_CLASS,
+      std::initializer_list<uint32_t>{uint32_t(SpvStorageClassFunction)});
+  in_operands.emplace_back(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                           std::initializer_list<uint32_t>{uint32_t(typeId)});
+  std::unique_ptr<ir::Instruction> type_inst(
+      new ir::Instruction(SpvOpTypePointer, 0, resultId, in_operands));
+  module_->AddType(std::move(type_inst));
+  return resultId;
+}
+
+void InlinePass::AddBranch(uint32_t labelId,
+                           std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> branch_in_operands;
+  branch_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{labelId}));
+  std::unique_ptr<ir::Instruction> newBranch(
+      new ir::Instruction(SpvOpBranch, 0, 0, branch_in_operands));
+  (*bp)->AddInstruction(std::move(newBranch));
+}
+
+void InlinePass::AddStore(uint32_t ptrId, uint32_t valId,
+                          std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> store_in_operands;
+  store_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{ptrId}));
+  store_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{valId}));
+  std::unique_ptr<ir::Instruction> newStore(
+      new ir::Instruction(SpvOpStore, 0, 0, store_in_operands));
+  (*bp)->AddInstruction(std::move(newStore));
+}
+
+void InlinePass::AddLoad(uint32_t typeId, uint32_t resultId, uint32_t ptrId,
+                         std::unique_ptr<ir::BasicBlock>* bp) {
+  std::vector<ir::Operand> load_in_operands;
+  load_in_operands.push_back(
+      ir::Operand(spv_operand_type_t::SPV_OPERAND_TYPE_ID,
+                  std::initializer_list<uint32_t>{ptrId}));
+  std::unique_ptr<ir::Instruction> newLoad(
+      new ir::Instruction(SpvOpLoad, typeId, resultId, load_in_operands));
+  (*bp)->AddInstruction(std::move(newLoad));
+}
+
+// Generate callee code into newBlocks to be inlined for the function call at
+// call_ii. Also add new function variables into caller func
+
+void InlinePass::GenInlineCode(
+    std::vector<std::unique_ptr<ir::BasicBlock>>* newBlocks,
+    std::vector<std::unique_ptr<ir::Instruction>>* newVars,
+    ir::UptrVectorIterator<ir::Instruction> call_ii,
+    ir::UptrVectorIterator<ir::BasicBlock> call_bi) {
+  // Map from callee id to caller id
+  std::unordered_map<uint32_t, uint32_t> callee2caller;
+  // Pre-call OpSampledImage Insts
```

Comment:
I sympathize that this causes complexity in this algorithm.

The rules are deliberate.  Checking the SPIR-V spec changelog, I refer you to Khronos private Bugzilla bugs 14241 and 14683.  (I believe you have access).

Whenever there is a tradeoff in complexity between a SPIR-V generator or consumer, the tendency is to make the consumer simpler.  The consumer is typically an API implementation (driver), which helps avoid bugs and improve consistency across the ecosystem.

Regarding how to handle this, I think we're very much helped by the fact that the OpSampledImage and OpImage instructions are always in the same block.  So sequential order also implies dominance order.  So I think you can (a) keep *both* OpSampledImage and OpImage instructions in the same preCall and postCall collections, and keep the old-id to new-idm mapping up to date while you are rewriting them in the multi-block case.
(We'd need tests for this too.)